src/util.rs

```diff
@@ -63,44 +65,114 @@ impl DebugVisualizations {
     ///
     /// # Panics
     /// Any failures in the conversion to CFG or RVSDG will cause a panic.
-    #[allow(unused)]
-    pub(crate) fn new(input: &str) -> DebugVisualizations {
+    pub fn new(input: &str) -> DebugVisualizations {
         let program = parse_from_string(input);
-        let mut cfg = to_cfg(&program.functions[0]);
-        let input_cfg = format!("{:#?}", Dot::new(&cfg.graph));
-        let restructured_cfg = {
-            let mut cfg = cfg.clone();
-            cfg.restructure();
-            format!("{:#?}", Dot::new(&cfg.graph))
-        };
-        let rvsdg = to_rvsdg(&mut cfg).unwrap().to_svg();
+        let cfg = Optimizer::program_to_cfg(&program);
+
+        let input_cfgs = cfg
+            .functions
+            .iter()
+            .map(|f| (f.name.clone(), format!("{:#?}", Dot::new(&f.graph))))
+            .collect();
+        let restructured_cfgs = cfg
+            .functions
+            .iter()
+            .map(|f| {
+                let mut f = f.clone();
+                f.restructure();
+                (f.name.clone(), format!("{:#?}", Dot::new(&f.graph)))
+            })
+            .collect();
+
+        let rvsdg = Optimizer::program_to_rvsdg(&program).unwrap().to_svg();
         DebugVisualizations {
-            input_cfg,
-            restructured_cfg,
+            input_cfgs,
+            restructured_cfgs,
             rvsdg,
         }
     }
 
-    /// Write the visualizations to output files with the given prefix.
+    /// Write the visualizations to output files in the given directory.
+    /// If the directory exists, it deletes it compeltely before writing.
+    /// Otherwise, it creates the directory (recursively).
     ///
-    /// * The `input_cfg` field is written to `prefix` + `input_cfg.dot`.
-    /// * The `restructured_cfg` field is written to `prefix` + `restructured.dot`.
-    /// * The `rvsdg` field is written to `prefix` + `rvsdg.svg`.
+    /// * The `input_cfg` field is written to `input_cfg.dot`.
+    /// * The `restructured_cfg` field is written to `function_name` + `restructured.dot`.
+    /// * The `rvsdg` field is written to `rvsdg.svg`.
     ///
     /// Like other utilities related to `DebugVisualizations`, this method is
     /// only intended for debugging eggcc.
-    #[allow(unused)]
-    pub(crate) fn write_output(&self, prefix: impl AsRef<str>) -> io::Result<()> {
+    pub fn write_output(&self, path: PathBuf) -> io::Result<()> {
         use std::fs::File;
         use std::io::Write;
-        for (name, content) in &[
-            ("input_cfg.dot", &self.input_cfg),
-            ("restructured_cfg.dot", &self.restructured_cfg),
-            ("rvsdg.svg", &self.rvsdg),
-        ] {
-            let mut file = File::create(format!("{}{}", prefix.as_ref(), name))?;
+
+        // make the directory, clearing it if needed
```

Comment:
I think this is dangerous / unexpected behavior. I'd expect this functionality to just overwrite anything in the supplied directory with a colliding name, not to clear the directory of all its contents (maybe I want to just dump the files into my working directory? but then it'd remove everything in there).

Reply:
yikes you are totally right! I'll put the prefix thingy back in as well